# Customer Segmentation
## Milestone 1 - Data Understanding

In [3]:
import pandas as pd 
from pathlib import Path 
PROJECT_ROOT = Path.cwd().parent
DATA_PATH = PROJECT_ROOT/"data"/"raw"/"online_retail_II.xlsx"


In [11]:
df_2010 = pd.read_excel(DATA_PATH, sheet_name = "Year 2009-2010")
df_2011 = pd.read_excel(DATA_PATH, sheet_name = "Year 2010-2011")
df = pd.concat([df_2010, df_2011], ignore_index=True)

In [5]:
df.head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.0,United Kingdom
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.0,United Kingdom
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.0,United Kingdom


In [6]:
df.shape

(1067371, 8)

In [7]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1067371 entries, 0 to 1067370
Data columns (total 8 columns):
 #   Column       Non-Null Count    Dtype         
---  ------       --------------    -----         
 0   Invoice      1067371 non-null  object        
 1   StockCode    1067371 non-null  object        
 2   Description  1062989 non-null  object        
 3   Quantity     1067371 non-null  int64         
 4   InvoiceDate  1067371 non-null  datetime64[us]
 5   Price        1067371 non-null  float64       
 6   Customer ID  824364 non-null   float64       
 7   Country      1067371 non-null  str           
dtypes: datetime64[us](1), float64(2), int64(1), object(3), str(1)
memory usage: 65.1+ MB


In [8]:
df.describe()

,Quantity,InvoiceDate,Price,Customer ID
count,1.067371e+06,1067371,1.067371e+06,824364.000000
mean,9.938898e+00,2011-01-02 21:13:55.394029,4.649388e+00,15324.638504
min,-8.099500e+04,2009-12-01 07:45:00,-5.359436e+04,12346.000000
25%,1.000000e+00,2010-07-09 09:46:00,1.250000e+00,13975.000000
50%,3.000000e+00,2010-12-07 15:28:00,2.100000e+00,15255.000000
75%,1.000000e+01,2011-07-22 10:23:00,4.150000e+00,16797.000000
max,8.099500e+04,2011-12-09 12:50:00,3.897000e+04,18287.000000
std,1.727058e+02,NaN,1.235531e+02,1697.464450


In [9]:
df.describe(include="object")

C:\Users\HP\AppData\Local\Temp\ipykernel_16608\702825166.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  df.describe(include="object")


,Invoice,StockCode,Description,Country
count,1067371,1067371,1062989,1067371
unique,53628,5305,5698,43
top,537434,85123A,WHITE HANGING HEART T-LIGHT HOLDER,United Kingdom
freq,1350,5829,5918,981330


In [10]:
df.isna().sum()

Invoice             0
StockCode           0
Description      4382
Quantity            0
InvoiceDate         0
Price               0
Customer ID    243007
Country             0
dtype: int64

# Updated Initial Observations
The dataset contains 1,067,371 transaction records across 8 columns.
Each row represents a transaction (or purchased item) rather than a unique customer. Customer-level features will need to be engineered before clustering.
The dataset spans from December 2009 to December 2011, providing approximately two years of transaction history.
The Customer ID column contains 243,007 missing values, which will require investigation during data cleaning.
The Description column contains only 4,382 missing values, while the remaining columns have no missing values.
The summary statistics show negative values in both Quantity and Price, indicating that there are special transactions (such as returns, cancellations, or corrections) that should be investigated before modeling.
The dataset contains 53,628 unique invoices, 5,305 unique stock codes, and transactions from 43 different countries, with the majority originating from the United Kingdom.

checking code to see if the dataset loaded correctly


In [14]:
import pandas as pd
from pathlib import Path

PROJECT_ROOT = PROJECT_ROOT = Path.cwd().parent
DATA_PATH = PROJECT_ROOT/"data"/"processed"/"customer_features.csv"

df = pd.read_csv(DATA_PATH)

In [15]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 5881 entries, 0 to 5880
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Customer ID        5881 non-null   float64
 1   Recency            5881 non-null   int64  
 2   Frequency          5881 non-null   int64  
 3   Monetary           5881 non-null   float64
 4   TotalQuantity      5881 non-null   int64  
 5   UniqueProducts     5881 non-null   int64  
 6   AverageOrderValue  5881 non-null   float64
dtypes: float64(3), int64(4)
memory usage: 321.7 KB
